src/array.cr

```diff
@@ -1045,6 +1026,64 @@ class Array(T)
     self
   end
 
+  # Inserts all of the elements from *other* before the element at *index*.
+  #
+  # This method shifts the element currently at *index* (if any) and any
+  # subsequent elements to the right, increasing their indices. If the value
+  # of *index* is negative, counting starts from the end of the array.
+  #
+  # Raises `IndexError` if the *index* is out of bounds.
+  #
+  # ```
+  # fruits = ["Apple"]
+  # newFruits = ["Dragonfruit", "Elderberry"]
+  #
+  # fruits.insert_all(1, newFruits)             # => ["Apple", "Dragonfruit", "Elderberry"]
+  # fruits.insert_all(-3, ["Banana", "Cherry"]) # => ["Apple", "Banana", "Cherry", "Dragonfruit", "Elderberry"]
+  #
+  # fruits.insert_all(6, ["invalid"])  # raises IndexError
+  # fruits.insert_all(-7, ["indices"]) # raises IndexError
+  # ```
+  def insert_all(index : Int, other : Enumerable(T) | Iterable(T)) : self
```

Comment:
**issue:** Allowing arbitrary `Enumerable` is problematic because the default implementation of `Enumerable#size` consumes the entire collection. `Iterable` doesn't even have a `#size` method.
So I think the type restriction should better be `Indexable`. And the generic argument should not be required to be `T`. This should be the same as the parameter for `#concat`. `#concat(other)` is semantically equivalent to `insert_all(-1, other)`.

Reply:
Oh, sorry. I hadn't noticed `#concat(Enumerable)` existed. However, I don't think this implementation could easily be adapted for generic `insert_all`. It would probably require multiple shifts of all array items after the insert index because the move distance is unknown. I reckon it would probably be most efficient to collect the enumerable into an array and insert that. There's no need to have an overload for this, an explicit `to_a` call would be sufficient.

For a first implementation, I'd restrict the type to `Indexable`.

If we notice a need for an optimized implementation for `Enumerable`, we can talk again. But I don't expect that to be much optimizable, except maybe when inserting close to the end of the array where you could pop a few elements and push them again after concatenating the enumerable.